The products data in the cart are loaded again from the database on every request which is made in the shop

acris-cp opened the pull request:
### 1. Why is this change necessary?
The products data in the cart are loaded again from the database on every request which is made in the shop (e.g. search, open a simple shop page, open a product page, ...)

### 2. What does this change do, exactly?
Optimisation of the function `isComplete` in the file `src/core/Content/Product/Cart/ProductCartProcessor.php` and some other things which have to be done according to this.

### 3. Describe each step to reproduce the issue or behaviour.
The problem is mainly due to the function `isComplete` in `src/core/Content/Product/Cart/ProductCartProcessor.php`.

There it asks whether the description of a product line item is null. However, this is ALWAYS zero, as it is never set for a product line item!
This has the consequence that Shopware reloads all products of the shopping basket in the line `$products = $this->productGateway->get($ids, $context);` in the method `collect` in the file mentioned above with EVERY request (i.e. also with the search, in the account area, every shop page, etc.).

This is exactly the opposite of what the new shopping cart was actually built for. An improvement brings a performance boost to every request made.

### 4. Please link to the relevant issues (if any).
https://issues.shopware.com/issues/NEXT-13250

### 5. Checklist

- [ ] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/master/adr/2020-08-03-Implement-New-Changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

aragon999 commented:
I am not sure if this function is needed, since it can be easily implemented but if, it should get implemented, I think the the container parameter `active_plugins` should be used to check if the required plugins are active, to avoid database queries:
```php
protected function assertRequiredPluginsPresent(array $plugins)
{
    return !(bool) array_diff(
        $plugins,
        array_keys($this->container->getParameter('active_plugins'))
    );
}
``` 
Maybe if such a function is implemented there are still some more things to be considered, the first to things which came into my mind are: 
- Check for the correct versions of each plugin
- Maybe it would be better not to return a Boolean value, but an array containing all the missing plugins

mitelg commented:
hey @Mardl 
thanks for your first PR! 
In the new plugin system the required plugins should be declared in the `plugin.xml` file of your plugin. 
See the XSD file https://github.com/shopware/shopware/blob/5.4/engine/Shopware/Components/Plugin/schema/plugin.xsd#L17

You can find an example here: https://developers.shopware.com/developers-guide/plugin-quick-start/#require-other-plugins

acris-cp committed:
Prevent loading cart items without changes

The main problem was, that in the function "isComplete" it asks for the product line item description and cover. But a description has never been set for a product line item. That means, it always returns false.

acris-cp committed:
Add flag modified to line items existing in cart

acris-cp committed:
Create  2021-01-13-improve-performance-loading-cart.md

J-Rahe commented:
Hey @acris-cp, yould you please take a look at the failing test and the csfixer and add some tests.

philipreinken commented:
Hello, we're closing this pull request due to inactivity. If this change is still important to you, feel free to create a new pull request.

For more information about our contribution guidelines, see https://docs.shopware.com/en/shopware-platform-dev-en/contribution/contribution-guideline.
